### PR TITLE
Add `--debug-rewrite` options

### LINKED ISSUE #2404: Debugging for rewrite rules
Just as there is an option `--debug-equation` to see when equations are attempted and why they are (or are not) applied, there should be a `--debug-rewrite` option to see when named rewrite rules are attempted and applied. This option would affect both `kprove` and `krun`.

### PR DESCRIPTION
**Pull request by alexd1971**
Fixes #2404 

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/Log/DebugAppliedRewriteRules.hs [hunk 1]
```diff
@@ -32,8 +33,9 @@ import Pretty (
 import Pretty qualified
 
 data DebugAppliedRewriteRules = DebugAppliedRewriteRules
-    { configuration :: Pattern VariableName
-    , appliedRewriteRules :: [SourceLocation]
+    { configuration :: !(Pattern VariableName)
+    , ruleLabels :: ![Text]
+    , appliedRewriteRules :: ![SourceLocation]
```
Comment (ana-pantilie) [inline kore/src/Kore/Log/DebugAppliedRewriteRules.hs:38]: Wouldn't this be a sum type of lables and sourcelocation? Basically, it should be `ruleId :: RuleId` where `RuleId = Label | SourceLocation`. Does that make sense or is there some detail I'm overlooking?

Comment (alexd1971) [reply kore/src/Kore/Log/DebugAppliedRewriteRules.hs:38]: I think it doesn't fit. Label and source location are used both. Label is used for selecting which debug message should be printed. And source location is used to print where the rule is placed in the source definition, like this:
```
The rules at following locations were applied:
 /home/alexey/Projects/RV/imp/imp.k:41:20-41:91
```

Comment (ana-pantilie) [reply kore/src/Kore/Log/DebugAppliedRewriteRules.hs:38]: Hmm, I think labels should be preferred over source locations. If a rule has a label, print the label, if it doesn't, print the source location

Comment (alexd1971) [reply kore/src/Kore/Log/DebugAppliedRewriteRules.hs:38]: Here is the output for option `--debug-attempt-equation`:
```
kore-exec: [237509] Debug (DebugAttemptEquation):
    applying equation at /home/alexey/Projects/RV/equational-rules/functions.k:21:20-21:59 to term:
        /* Fl Fn D Spa */
        Lbllength'LParUndsRParUnds'FUNCTIONS'Unds'Int'Unds'ConsList{}(
            /* Fl Fn D Sfa */ RuleVarLIST:SortConsList{}
        )
Context:
    (DebugAttemptEquation) while applying equation at /home/alexey/Projects/RV/equational-rules/functions.k:21:20-21:59
kore-exec: [239006] Debug (DebugAttemptEquation):
    equation is not applicable:
    equation did not match term:  unimplemented matching case
Context:
    (DebugAttemptEquation) while applying equation at /home/alexey/Projects/RV/equational-rules/functions.k:21:20-21:59
kore-exec: [245011] Debug (DebugAttemptEquation):
    applying equation at /home/alexey/Projects/RV/equational-rules/functions.k:21:20-21:59 to term:
        /* Fl Fn D Spa */
        Lbllength'LParUndsRParUnds'FUNCTIONS'Unds'Int'Unds'ConsList{}(
            /* Fl Fn D Sfa */ ConfigVarLIST:SortConsList{}
        )
Context:
    (InfoReachability) while simplifying the claim
        (DebugAttemptEquation) while applying equation at /home/alexey/Projects/RV/equational-rules/functions.k:21:20-21:59
kore-exec: [246091] Debug (DebugAttemptEquation):
    equation is not applicable:
    equation did not match term:  unimplemented matching case
Context:
    (DebugAttemptEquation) while applying equation at /home/alexey/Projects/RV/equational-rules/functions.k:21:20-21:59
        (InfoReachability) while simplifying the claim
#Top
```
Here we see source location for every attempt, and don't see any label information. Label is used only to select which debug messages should be printed. In similar way we get debug output for `--debug-attempt-rewrite` option:
```
kore-exec: [330673] Debug (DebugAttemptedRewriteRules):
    The rule at following location was attempted:
    /home/alexey/Projects/RV/imp/imp.k:113:21-113:73
    On configuration:
      \and{SortGeneratedTopCell{}}(
          /* term: */
          /* Fl Fn D Sfa */
          Lbl'-LT-'generatedTop'-GT-'{}(
              /* Fl Fn D Sfa Cl */
              Lbl'-LT-'T'-GT-'{}(
                  /* Fl Fn D Sfa Cl */
                  Lbl'-LT-'k'-GT-'{}(
                      /* Fl Fn D Sfa Cl */
                      kseq{}(
...
```
My changes doesn't affect the debug output format. I only added options an its selection by label. Therefore I think that if we want to change the debug output and to add to it the selected label, then we need to open a separate issue for this.

Comment (ana-pantilie) [reply kore/src/Kore/Log/DebugAppliedRewriteRules.hs:38]: Ok, thank you for investigating. Will do!

**Review by ana-pantilie — APPROVED** [review]
👍

**Pull request merged**